ykcompile/src/lib.rs

```diff
@@ -473,160 +485,99 @@ impl<TT> TraceCompiler<TT> {
                     );
                 }
                 4 => {
                     dynasm!(self.asm
-                        ; mov DWORD [Rq(ro.reg) + ro.offs], Rd(reg)
+                        ; mov Rd(reg), [Rq(ro.reg) + ro.offs]
                     );
                 }
                 8 => {
                     dynasm!(self.asm
-                        ; mov [Rq(ro.reg) + ro.offs], Rq(reg)
+                        ; mov Rq(reg), [Rq(ro.reg) + ro.offs]
                     );
                 }
                 _ => todo!("{}", ty.size()),
             },
-            (Location::Mem(ro1), Location::Mem(ro2)) => {
-                // Since RAX is currently not available to the register allocator, we can use it
-                // here to simplify moving values from the stack back onto the stack (which x86
-                // does not support). Otherwise, we would have to free up a register via spilling,
-                // making this operation more complicated and costly.
-                match ty.size() {
-                    1 => {
-                        dynasm!(self.asm
-                            ; mov al, BYTE [Rq(ro2.reg) + ro2.offs]
-                            ; mov BYTE [Rq(ro1.reg) + ro1.offs], al
-                        );
-                    }
-                    8 => {
-                        dynasm!(self.asm
-                            ; mov rax, [Rq(ro2.reg) + ro2.offs]
-                            ; mov [Rq(ro1.reg) + ro1.offs], rax
-                        );
-                    }
-                    16 => {
-                        // We could have handled this case via a generic memcpy() solution, but it
-                        // hardly seems worth the overhead for such a small copy. Especially since
-                        // pairs (tuples) of two u64s are very common in Rust: they are the result
-                        // of checked arithmetic.
-                        dynasm!(self.asm
-                            ; mov rax, [Rq(ro2.reg) + ro2.offs]
-                            ; mov [Rq(ro1.reg) + ro1.offs], rax
-                            ; mov rax, [Rq(ro2.reg) + ro2.offs + 8]
-                            ; mov [Rq(ro1.reg) + ro1.offs + 8], rax
-                        );
-                    }
-                    _ => todo!("{}", ty.size()), // FIXME: For things >16, use memcpy().
-                }
-            }
             _ => unreachable!(),
         }
-
-        // Free temporary if one was created.
-        self.free_if_temp(lloc);
-        self.free_if_temp(rloc);
+        self.free_if_temp(loc);
     }
 
-    fn mov_place_ref(&mut self, p1: &Place, p2: &Place) {
-        let (lloc, _) = self.place_to_location(p1);
-
+    /// Codegen a reference `p` and store the result in `reg`.
+    fn c_ref(&mut self, reg: u8, p: &Place) {
         // Deal with the special case `&*`, i.e. referencing a `Deref` on a reference just returns
         // the reference.
```

Comment:
I find this a bit hard to read so I have a dumb question. Will this work with multiple deref operators: `&**`, e.g.:
```
let s = &String::new("hello");
let str_ = &**s;
```
I _think_ `str_` should be of type `&str` here, because we first deref the `&String` to get `String`, and then `&*` will convert it to `&str` because it will call `String::deref` (where the deref method here does implicit conversion).

Reply:
Apart from the fact that we can't deal with custom deref implementations anyway yet, there is still a bug here. Your example shouldn't hit the special case, which is limited to `&*` on `Ref`s. So it should instead hit the bottom case which would just resolve the projections in the correct order. However, because our conditions are a bit wonky, your example would actually hit the special case which is wrong and needs to be fixed. But that belongs in a separate PR.

Reply:
Ok I see. Perhaps add a comment that summarises this in a FIXME?

Reply:
> Apart from the fact that we can't deal with custom deref implementations anyway yet

Those are lowered into calls when building MIR.

Reply:
Right, I've already forgotten about that. :-)

Reply:
Can we FIXME here before merging?